src/main/java/com/commercetools/sync/products/ProductSync.java

```diff
@@ -181,11 +181,16 @@ private void processFetchedProducts(@Nonnull final Set<Product> fetchedProducts)
     }
 
     @Nonnull
-    private CompletionStage<List<Optional<Product>>> createOrUpdateProducts() {
-        return productService.createProducts(draftsToCreate)
-                             .thenAccept(createdProducts ->
-                                 updateStatistics(createdProducts, draftsToCreate.size()))
-                             .thenCompose(ignoredResult -> syncProducts(productsToSync));
+    private CompletionStage<Void> createOrUpdateProducts() {
+        final CompletableFuture<Void> futureCreates =
+            productService.createProducts(draftsToCreate)
+                          .thenAccept(createdProducts -> updateStatistics(createdProducts, draftsToCreate.size()))
+                          .toCompletableFuture();
+
+        final CompletableFuture<List<Optional<Product>>> futureUpdates =
+            syncProducts(productsToSync).toCompletableFuture();
+
+        return CompletableFuture.allOf(futureCreates, futureUpdates);
```

Comment:
For the whole approach in this function i'd say `CompletableFuture#thenCombine()` looks simpler and more sufficient, than `CompletableFuture#allOf()`:
```java
return productService.createProducts(draftsToCreate)
    .thenAccept(createdProducts -> updateStatistics(createdProducts, draftsToCreate.size()))
    .thenCombine(syncProducts(productsToSync), (void1, void2) -> null);
```

Reply:
2e15fe9